File: Basics/Tree Traversal/Tree.go
package main

import "fmt"

type TreeNode struct {
	value       int
	left, right *TreeNode
}

type BinaryTree struct {
	root *TreeNode
}

func NewTreeNode(value int) *TreeNode {
	return &TreeNode{value, nil, nil}
}

func NewBinaryTree() *BinaryTree {
	return &BinaryTree{nil}
}

func (bt *BinaryTree) Insert(value int) {
	bt.root = bt.insert(bt.root, value)
}

func (bt *BinaryTree) insert(root *TreeNode, value int) *TreeNode {
	if root == nil {
		return NewTreeNode(value)
	}

	if value < root.value {
		root.left = bt.insert(root.left, value)
	} else {
		root.right = bt.insert(root.right, value)
	}

	return root
}

func (bt *BinaryTree) InOrderTraversal() []int {
	result := []int{}
	bt.inOrderTraversal(bt.root, &result)
	return result
}

func (bt *BinaryTree) inOrderTraversal(node *TreeNode, result *[]int) {
	if node != nil {
		bt.inOrderTraversal(node.left, result)
		*result = append(*result, node.value)
		bt.inOrderTraversal(node.right, result)
	}
}

func (bt *BinaryTree) PreOrderTraversal() []int {
	result := []int{}
	bt.preOrderTraversal(bt.root, &result)
	return result
}

func (bt *BinaryTree) preOrderTraversal(node *TreeNode, result *[]int) {
	if node != nil {
		*result = append(*result, node.value)
		bt.preOrderTraversal(node.left, result)
		bt.preOrderTraversal(node.right, result)
	}
}

func (bt *BinaryTree) PostOrderTraversal() []int {
	result := []int{}
	bt.postOrderTraversal(bt.root, &result)
	return result
}

func (bt *BinaryTree) postOrderTraversal(node *TreeNode, result *[]int) {
	if node != nil {
		bt.postOrderTraversal(node.left, result)
		bt.postOrderTraversal(node.right, result)
		*result = append(*result, node.value)
	}
}

// Example usage:
func main() {
	myTree := NewBinaryTree()
	myTree.Insert(5)
	myTree.Insert(3)
	myTree.Insert(7)
	myTree.Insert(2)
	myTree.Insert(4)
	myTree.Insert(6)
	myTree.Insert(8)

	fmt.Println("In-order traversal:", myTree.InOrderTraversal())
	fmt.Println("Pre-order traversal:", myTree.PreOrderTraversal())
	fmt.Println("Post-order traversal:", myTree.PostOrderTraversal())
}

File: Basics/Tree Traversal/Tree.rs
use std::fmt;

#[derive(Debug)]
struct TreeNode {
    value: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(value: i32) -> Self {
        TreeNode {
            value,
            left: None,
            right: None,
        }
    }
}

#[derive(Debug)]
struct BinaryTree {
    root: Option<Box<TreeNode>>,
}

impl BinaryTree {
    fn new() -> Self {
        BinaryTree { root: None }
    }

    fn insert(&mut self, value: i32) {
        self.root = self._insert(self.root.take(), value);
    }

    fn _insert(&mut self, node: Option<Box<TreeNode>>, value: i32) -> Option<Box<TreeNode>> {
        match node {
            None => Some(Box::new(TreeNode::new(value))),
            Some(mut n) => {
                if value < n.value {
                    n.left = self._insert(n.left.take(), value);
                } else {
                    n.right = self._insert(n.right.take(), value);
                }
                Some(n)
            }
        }
    }

    fn in_order_traversal(&self) -> Vec<i32> {
        let mut result = Vec::new();
        self._in_order_traversal(self.root.as_ref(), &mut result);
        result
    }

    fn _in_order_traversal(&self, node: Option<&Box<TreeNode>>, result: &mut Vec<i32>) {
        if let Some(n) = node {
            self._in_order_traversal(n.left.as_ref(), result);
            result.push(n.value);
            self._in_order_traversal(n.right.as_ref(), result);
        }
    }

    fn pre_order_traversal(&self) -> Vec<i32> {
        let mut result = Vec::new();
        self._pre_order_traversal(self.root.as_ref(), &mut result);
        result
    }

    fn _pre_order_traversal(&self, node: Option<&Box<TreeNode>>, result: &mut Vec<i32>) {
        if let Some(n) = node {
            result.push(n.value);
            self._pre_order_traversal(n.left.as_ref(), result);
            self._pre_order_traversal(n.right.as_ref(), result);
        }
    }

    fn post_order_traversal(&self) -> Vec<i32> {
        let mut result = Vec::new();
        self._post_order_traversal(self.root.as_ref(), &mut result);
        result
    }

    fn _post_order_traversal(&self, node: Option<&Box<TreeNode>>, result: &mut Vec<i32>) {
        if let Some(n) = node {
            self._post_order_traversal(n.left.as_ref(), result);
            self._post_order_traversal(n.right.as_ref(), result);
            result.push(n.value);
        }
    }
}

// Example usage:
fn main() {
    let mut my_tree = BinaryTree::new();
    my_tree.insert(5);
    my_tree.insert(3);
    my_tree.insert(7);
    my_tree.insert(2);
    my_tree.insert(4);
    my_tree.insert(6);
    my_tree.insert(8);

    println!("In-order traversal: {:?}", my_tree.in_order_traversal());
    println!("Pre-order traversal: {:?}", my_tree.pre_order_traversal());
    println!("Post-order traversal: {:?}", my_tree.post_order_traversal());
}
